### ISSUE TITLE
Visual Studio is busy hang with no resource usage

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#56988]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Visual-Studio-is-busy-hang-with-no-resou/1507976)._

---
[severity:It's more difficult to complete my work]
I've been having issues related to IntelliSense or building the solution while working through a guide teaching unreal engine. I can't quite pinpoint if it is related to intellisense (which can take minutes sometimes to calculate and start coloring elements or doing much of anything especially if I change a header file) or if it is from building (as it sometimes happens after building solution) but multiple times now I've had to end task from task manager because visual studio entirely hangs and won't respond in any way other than bringing the window to the front when the taskbar icon is clicked and putting a notification saying "Visual Studio is busy; We'll automatically report this delay to Microsoft based on your Visual Studio Experience Improvement Program settings." which is becoming quite frustrating to constantly reopen the project. I did generate a dump file of the main process from task manager that I will attach if it helps at all.

---
#### Feedback Bot on 8/19/2021, 01:43 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 8/23/2021, 08:55 PM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

**Comment by jmarolf** [comment]
This is a deadlock.

UI thread is blocked on a lock here:

```ini
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.RefreshOption(Microsoft.CodeAnalysis.Options.OptionKey optionKey, object newValue) Line 414	C#
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister.RefreshLanguageSettings(Microsoft.VisualStudio.TextManager.Interop.LANGPREFERENCES3[] langPrefs) Line 115	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister.LanguageSettingsPersister(Microsoft.CodeAnalysis.Editor.Shared.Utilities.IThreadingContext threadingContext, Microsoft.VisualStudio.TextManager.Interop.IVsTextManager4 textManager, Microsoft.CodeAnalysis.Options.IGlobalOptionService optionService) Line 64	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersisterProvider.GetOrCreatePersisterAsync(System.Threading.CancellationToken cancellationToken) Line 50	C#
```

That same lock is held by this thread pool worker, which itself has performed a `JTF.Run` around a `SwitchToMainThreadAsync`, trying to create above `LanguageSettingsPersisterProvider`:

```ini 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.WorkspaceThreadingService.Run<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Options.IOptionPersister>>(System.Func<System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Options.IOptionPersister>>> asyncMethod)	Unknown
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOptionPersisters.__GetOptionPersistersSlow|16_0(Microsoft.CodeAnalysis.Shared.Utilities.IWorkspaceThreadingService workspaceThreadingService, System.Collections.Immutable.ImmutableArray<System.Lazy<Microsoft.CodeAnalysis.Options.IOptionPersisterProvider>> optionSerializerProviders, System.Threading.CancellationToken cancellationToken) Line 124	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOptionPersisters() Line 114	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.LoadOptionFromSerializerOrGetDefault(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 144	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption_NoLock(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 345	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 326	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.OptionsHelpers.GetOption<bool>(Microsoft.CodeAnalysis.Options.OptionKey optionKey, System.Func<Microsoft.CodeAnalysis.Options.OptionKey, object> getOption) Line 29	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption<bool>(Microsoft.CodeAnalysis.Options.Option2<bool> option) Line 312	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.OptionServiceFactory.OptionService.GetOption<bool>(Microsoft.CodeAnalysis.Options.Option2<bool> option) Line 120	C#
```

NOTE: By inspection I see that every `IOptionPersisterProvider` switches to the UI thread, so the fix should probably to be not to hold a lock that the UI thread needs while creating these services.

**Comment by jmarolf** [comment]
Quoting analysis from @davkean:

This is a deadlock.

UI thread is blocked on a lock here:

```ini
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.RefreshOption(Microsoft.CodeAnalysis.Options.OptionKey optionKey, object newValue) Line 414	C#
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister.RefreshLanguageSettings(Microsoft.VisualStudio.TextManager.Interop.LANGPREFERENCES3[] langPrefs) Line 115	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister.LanguageSettingsPersister(Microsoft.CodeAnalysis.Editor.Shared.Utilities.IThreadingContext threadingContext, Microsoft.VisualStudio.TextManager.Interop.IVsTextManager4 textManager, Microsoft.CodeAnalysis.Options.IGlobalOptionService optionService) Line 64	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersisterProvider.GetOrCreatePersisterAsync(System.Threading.CancellationToken cancellationToken) Line 50	C#
```

That same lock is held by this thread pool worker, which itself has performed a JTF.Run around a `SwitchToMainThreadAsync, trying to create above LanguageSettingsPersisterProvider:

```ini 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.WorkspaceThreadingService.Run<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Options.IOptionPersister>>(System.Func<System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Options.IOptionPersister>>> asyncMethod)	Unknown
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOptionPersisters.__GetOptionPersistersSlow|16_0(Microsoft.CodeAnalysis.Shared.Utilities.IWorkspaceThreadingService workspaceThreadingService, System.Collections.Immutable.ImmutableArray<System.Lazy<Microsoft.CodeAnalysis.Options.IOptionPersisterProvider>> optionSerializerProviders, System.Threading.CancellationToken cancellationToken) Line 124	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOptionPersisters() Line 114	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.LoadOptionFromSerializerOrGetDefault(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 144	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption_NoLock(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 345	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption(Microsoft.CodeAnalysis.Options.OptionKey optionKey) Line 326	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.OptionsHelpers.GetOption<bool>(Microsoft.CodeAnalysis.Options.OptionKey optionKey, System.Func<Microsoft.CodeAnalysis.Options.OptionKey, object> getOption) Line 29	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption<bool>(Microsoft.CodeAnalysis.Options.Option2<bool> option) Line 312	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.OptionServiceFactory.OptionService.GetOption<bool>(Microsoft.CodeAnalysis.Options.Option2<bool> option) Line 120	C#
```

NOTE: By inspection I see that every IOptionPersisterProvider switches to the UI thread, so the fix should probably to be not to hold a lock that the UI thread needs while creating these services.

**Comment by sharwell** [comment]
@jmarolf @tmat This was already fixed by #54845
